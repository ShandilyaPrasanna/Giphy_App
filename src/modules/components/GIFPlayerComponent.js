import React, { Component } from "react";
import ReactDOM from "react-dom";

export class GIFPlayerComponent extends Component {
  render() {
    return ReactDOM.createPortal(
      <div className="popover">
        <div className="blurBackground" />
        <div id="player" className="player">
          <div id="closePlayer" onClick={this.props.closePlayer}>
            X
          </div>
          <video width="100%" height="100%" controls autoplay loop muted>
            <source src={this.props.GIFToPlay} type="video/mp4" />
          </video>
        </div>
      </div>,

      document.getElementById("GIFPlayer")
    );
  }
}

export default GIFPlayerComponent;
