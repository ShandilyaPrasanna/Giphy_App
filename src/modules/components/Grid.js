import React, { Component } from "react";
import { connect } from "react-redux";
import GIFPlayerComponent from "./GIFPlayerComponent";
import { getGIFSByKeywordAction, getTrendingGIFS, isLoading } from "../actions";

export class Grid extends Component {
  constructor(props) {
    super(props);
    this.state = {
      count: 25,
      openGIFsPlayer: false,
      GIFToPlay: {}
    };
    this.scrollListener = window.addEventListener("scroll", e => {
      this.handleScroll(e);
    });
  }

  handleScroll = () => {
    const { count } = this.state;
    const {
      getTrendingGIFS,
      getGIFSByKeywordAction,
      searchText,
      isLoading,
      loading
    } = this.props;
    const totalCount =
      this.props.data &&
      this.props.data.pagination &&
      this.props.data.pagination.total_count;

    if (loading && totalCount <= count) return;
    var lastLi = document.querySelector(".Grid > .imageDiv:last-child");
    var lastLiOffset = lastLi.offsetTop + lastLi.clientHeight;
    var pageOffset = window.pageYOffset + window.innerHeight;
    var bottomOffset = 20;
    if (pageOffset > lastLiOffset - bottomOffset) {
      this.setState({ count: count + 10 }, () => {
        searchText === ""
          ? getTrendingGIFS(this.state.count)
          : getGIFSByKeywordAction(searchText, this.state.count);
        isLoading(true);
      });
    }
  };

  closePlayer = () => {
    this.setState({ openGIFsPlayer: false, GIFToPlay: "" });
  };

  showGIFPlayer = url => {
    this.setState({ openGIFsPlayer: true, GIFToPlay: url });
  };

  render() {
    return (
      <React.Fragment>
        {this.state.openGIFsPlayer && (
          <GIFPlayerComponent
            GIFToPlay={this.state.GIFToPlay}
            closePlayer={this.closePlayer}
          />
        )}
        <div className="Grid">
          {this.props.data &&
            this.props.data.images &&
            this.props.data.images.map((ele, index) => {
              return (
                <div
                  key={ele._id}
                  className="imageDiv"
                  onClick={() => {
                    this.showGIFPlayer(ele.images.original.mp4);
                  }}
                >
                  <img
                    key={ele._id}
                    className="imageTag"
                    src={ele.images.original_still.url}
                    alt="Loading"
                  />
                </div>
              );
            })}
        </div>
      </React.Fragment>
    );
  }
}

export function mapStateToProps(state) {
  return {
    data: state.app,
    loading: state.app.loading
  };
}

export default connect(
  mapStateToProps,
  { getGIFSByKeywordAction, getTrendingGIFS, isLoading }
)(Grid);
