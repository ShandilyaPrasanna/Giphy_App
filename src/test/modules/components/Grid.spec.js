import React from "react";
import ReactDOM from "react-dom";
import renderer from "react-test-renderer";
import toJson from "enzyme-to-json";
import { shallow, configure } from "enzyme";
import Adapter from "enzyme-adapter-react-16";
import { Grid, mapStateToProps } from "../../../modules/components/Grid";

document.querySelector = jest.fn().mockImplementation(() => {
  return { offsetTop: 10, clientHeight: 100 };
});
const data = [
  {
    id: "7lD9Gz5FxpRCg",
    images: {
      fixed_height: {
        url:
          "https://media2.giphy.com/media/7lD9Gz5FxpRCg/giphy-fixed_height.gif",
        width: "500",
        height: "237",
        size: "1990441"
      },
      original_still: {
        url:
          "https://media2.giphy.com/media/7lD9Gz5FxpRCg/giphy-fixed_height.gif",
        width: "500",
        height: "237",
        size: "1990441"
      },
      original: {
        url:
          "https://media2.giphy.com/media/7lD9Gz5FxpRCg/giphy-fixed_height.gif",
        width: "500",
        height: "237",
        size: "1990441"
      }
    }
  }
];
const URL = "https://media1.giphy.com/media/aGyZ004XM4MjC/100w.gif";
const window = { pageYOffset: 20, innerHeight: 200 };

configure({ adapter: new Adapter() });
describe("Grid Test", () => {
  let tree;
  const props = {
    getGIFSByKeywordAction: jest.fn(),
    getTrendingGIFS: jest.fn(),
    isLoading: jest.fn(),
    data: { images: data, pagination: { total_count: 100 }, loading: false },
    loading: false
  };
  beforeEach(() => {
    tree = shallow(<Grid {...props} />);
  });

  it("renders Grid ", () => {
    expect(toJson(tree)).toMatchSnapshot();
  });

  it("onCLick should open popover with GIF ", () => {
    tree.find(".imageDiv").simulate("click", URL);
    expect(tree.state("openGIFsPlayer")).toBe(true);
  });

  it("test closePlayer callcack ", () => {
    tree.instance().closePlayer();
    expect(tree.state("openGIFsPlayer")).toBe(false);
  });

  it("test mapStateToProps ", () => {
    const state = {
      app: {
        loading: true,
        images: [],
        pagination: {}
      }
    };

    expect(mapStateToProps(state).loading).toBe(true);
  });

  it("test handleScroll function ", () => {
    tree.setProps({ loading: true });
    tree.setState({ count: 25 });
    tree.instance().handleScroll();
    expect(tree.state("count")).toBe(35);
  });
});
