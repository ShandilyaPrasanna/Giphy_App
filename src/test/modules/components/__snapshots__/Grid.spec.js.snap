// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Grid Test renders Grid  1`] = `
<Fragment>
  <div
    className="Grid"
  >
    <div
      className="imageDiv"
      onClick={[Function]}
    >
      <img
        alt="Loading"
        className="imageTag"
        src="https://media2.giphy.com/media/7lD9Gz5FxpRCg/giphy-fixed_height.gif"
      />
    </div>
  </div>
</Fragment>
`;
